### PR TITLE
Complemented low control

### PR DESCRIPTION
**Pull request by Anders-Holst**
This pull request contains the changes to the low-level control interface, ControlInterface, suggested in the unwieldy pull request #80 .

(I really tried to make the commit messages more specific this time. But with so many changes, some are still kind of vague.)

  / Anders

### CODE REVIEW

**Comment by Anders-Holst** [comment]
@scrool , I have now removed the realtime UDP protocols from this pull request, as you wished.

I am standing by to add it again, together with the new unit test, once you have merged this one.

And in line after that are all the other low-level unit tests.

**Comment by scrool** [comment]
@Anders-Holst I have noticed that you haven't run `black` linter on your changes. Usually it would be run automatically with pre-commit hook - see https://github.com/scrool/xled/blob/develop/CONTRIBUTING.rst how to set it up:

```
$ pip install -r requirements_dev.txt
$ pre-commit install
```

To not miss this next time I have updated GitHub actions to run linters for pull requests but this doesn't run against this PR because it is based on `develop` branch before I added #88.

Please:

1. rebase your changes against `develop`
2. make sure you run `black` and commit all its changes.

**Comment by Anders-Holst** [comment]
OK, you are right, I had missed black. Tried to rebase and apply black. I thought I had fixed it all now. (Obviously did a small mistake when committing, and git rebase seemed confused so I was not able to just "drop" my last commit.)
But when pushing I still get complaints from "lint". Why? I can not find any log saying what it complains on.
Getting very tired, need to sleep...

**Comment by scrool** [comment]
From what I can tell, `black` passes now. :+1: 

There is one error caught by `flake8` though:
```
xled/control.py:33:1: F401 'xled.security.encrypt_wifi_password' imported but unused
```

that's because you import directly method `encrypt_wifi_password` but later you use `xled.security.encrypt_wifi_password`.

**Comment by Anders-Holst** [comment]
New try. Fixed the problem, passes black and flake8 here. Hope it passes on your side too.
(I really tried to undo the two confused commits from last night, but the first seem to be "special" as it was the result of trying to solve conflicts appearing during rebase, so can't be easily removed. So I let it be, although the history looks ugly...)

File: xled/control.py [hunk 1]
```diff
@@ -314,72 +586,289 @@ def set_led_movie_config(self, frame_delay, frames_number, leds_number):
...
+
+    def set_rt_frame_rest(self, frame):
+        """
+        Uploads a frame in rt-mode, using the ordinary restful protocol
+
+        :param frame: file-like object that points to frame file.
+        :raises ApplicationError: on application error
+        :rtype: :class:`~xled.response.ApplicationResponse`
+        """
+        url = urljoin(self.base_url, "led/rt/frame")
```
Comment (magicus) [inline xled/control.py:833]: I'm sorry that I've missed much of the discussion here. This looks interesting! Is it a change with new firmware versions to support this? How did you find out this call existed?

**Review by magicus — APPROVED** [review]
From what I can tell, this looks like a very good change.

File: xled/control.py (hunk 1 above)
Comment (scrool) [reply xled/control.py:833]: @magicus see https://xled-docs.readthedocs.io/en/latest/rest_api.html#send-realtime-frame . I found that https://github.com/scrool/xled-docs/commit/4dc705574ac166f10b1bf7e89a1c50a86f163371 , I believe by watching data on a wire.

**Comment by scrool** [comment]
I have skipped those commits which didn't provide any changes and merged the rest into `develop`.

Thanks again! :partying_face:

File: xled/control.py (hunk 1 above)
Comment (magicus) [reply xled/control.py:833]: I see, thanks. I have not been following the updates to the docs.

**Pull request closed without merging**